lib/datadog/core/configuration/option.rb

```diff
@@ -9,17 +9,36 @@ class Option
         attr_reader \
           :definition
 
+        # Option setting precedence. Higher number means higher precedence.
+        module Precedence
+          REMOTE_CONFIGURATION = 2
+          PROGRAMMATIC = 1
+          DEFAULT = 0
+        end
+
         def initialize(definition, context)
           @definition = definition
           @context = context
           @value = nil
           @is_set = false
+
+          # Lowest precedence, to allow for `#set` to always succeed for a brand new `Option` instance.
+          @precedence_set = Precedence::DEFAULT
         end
 
-        def set(value)
+        # Overrides the current value for this option if the `precedence` is equal or higher than
+        # the previously set value.
+        # The first call to `#set` will always store the value regardless of precedence.
+        #
+        # @param value [Object] the new value to be associated with this option
+        # @param precedence [Precedence] from what precedence order this new value comes from
+        def set(value, precedence: Precedence::PROGRAMMATIC)
+          return @value if precedence < @precedence_set # Cannot override higher precedence value
```

Comment:
On the topic of warnings when there are clashes between settings (comment above), I think we should at least print a warning when a setting gets ignored here.

E.g. it's very counter-intuitive that I would run a `Datadog.configure { }` block 10 minutes after I started my application and my setting would get outright ignored because in the meanwhile it got set via remote configuration.

Reply:
Done!